libass/ass_render.c

```diff
@@ -2203,244 +2134,192 @@ static void render_and_combine_glyphs(ASS_Renderer *render_priv,
     CombinedBitmapInfo *combined_info = text_info->combined_bitmaps;
     CombinedBitmapInfo *current_info = NULL;
     GlyphInfo *last_info = NULL;
-    for (i = 0; i < text_info->length; ++i) {
+    for (int i = 0; i < text_info->length; ++i) {
         GlyphInfo *info = text_info->glyphs + i;
         if (info->linebreak) linebreak = 1;
         if (info->skip) continue;
-        while (info) {
+        for (; info; info = info->next) {
             OutlineBitmapHashKey *key = &info->hash_key.u.outline;
-            info->pos.x *= render_priv->font_scale_x;
-            key->advance.x =
-                double_to_d6(device_x - (int) device_x +
-                        d6_to_double(info->pos.x & SUBPIXEL_MASK)) & ~SUBPIXEL_ACCURACY;
-            key->advance.y =
-                double_to_d6(device_y - (int) device_y +
-                        d6_to_double(info->pos.y & SUBPIXEL_MASK)) & ~SUBPIXEL_ACCURACY;
-            get_bitmap_glyph(render_priv, info);
 
-            int bm_x = info->pos.x >> 6,
-                bm_y = info->pos.y >> 6,
-                bm_o_x = bm_x, bm_o_y = bm_y, min_bm_x = bm_x, min_bm_y = bm_y;
-
-            if(info->bm){
-                bm_x += info->bm->left;
-                bm_y += info->bm->top;
-                min_bm_x = bm_x;
-                min_bm_y = bm_y;
-            }
-
-            if(info->bm_o){
-                bm_o_x += info->bm_o->left;
-                bm_o_y += info->bm_o->top;
-                min_bm_x = FFMIN(min_bm_x, bm_o_x);
-                min_bm_y = FFMIN(min_bm_y, bm_o_y);
-            }
+            info->pos.x = double_to_d6(device_x + d6_to_double(info->pos.x) * render_priv->font_scale_x);
+            info->pos.y = double_to_d6(device_y) + info->pos.y;
+            key->advance.x = info->pos.x & (SUBPIXEL_MASK & ~SUBPIXEL_ACCURACY);
+            key->advance.y = info->pos.y & (SUBPIXEL_MASK & ~SUBPIXEL_ACCURACY);
+            int x = info->pos.x >> 6, y = info->pos.y >> 6;
+            get_bitmap_glyph(render_priv, info);
+            if (!info->image)
+                continue;
```

Comment:
This conditional should come after the following one, to ensure that we correctly split runs even when we fail to rasterize some glyphs.